Deduplicate a bunch of flasher progress code

arkeet opened the pull request:
Follow-up to discussion in #3269. This removes a lot of flasher progress code that is duplicated for each operation (which became even worse now that I added one). This brings in [`enum_map`](https://crates.io/crates/enum-map/2.7.3) as a dependency, enabling type-safe mappings indexed by `ProgressOperation`. It should functionally be a no-op.

Yatekii commented on probe-rs-tools/Cargo.toml:
```diff
@@ -38,6 +38,7 @@ thiserror = { workspace = true }
 tracing = "0.1"
 typed-path = "0.10"
 probe-rs-mi = { workspace = true }
+enum-map = "2.7.3"
```
```suggestion
enum-map = "2.7"
```

We typically don't include the patch version :) It will take the latest anyways.

Yatekii reviewed (approved):
Amazing change, thanks a lot!

bugadani reviewed (changes requested):
Sorry to step in I just want to take a look before this gets merged

bugadani commented on probe-rs-tools/src/bin/probe-rs/rpc/functions/flash.rs:
```diff
@@ -204,6 +204,17 @@ impl From<flashing::ProgressOperation> for Operation {
     }
 }
 
+impl From<Operation> for flashing::ProgressOperation {
```
I'd prefer not adding this. If you stick with enum_map, the probe-rs-tools types should implement it, and the library API itself should be left unchanged.

bugadani replied:
I refer to `derive(enum_map::Enum)` in the library.

arkeet replied:
Understandable. I was hesitant to add the `Enum` impl to the public library type; I think at one point I had something using `EnumMap` in there but it is gone now. I'll move it to the tools type.

Do you have an opinion on the new `FlashProgress` methods? I was thinking eventually the 4×4 methods for each combination of event and operation could be inlined and removed, but this change perhaps awkwardly goes only halfway. (My original motivation to add them was that my previous PR shared code between preverify and verify, with the only difference being the `ProgressOperation` value.)

bugadani replied:
You mean the `pub(super) fn progressed(` stuff? I like it, its cleaner than I thought it would be.

arkeet commented on probe-rs-tools/src/bin/probe-rs/cmd/dap_server/server/debugger.rs:
```diff
@@ -581,165 +582,68 @@ impl Debugger {
+                    ProgressEvent::AddProgressBar { operation, total } => {
+                        if let Some(total) = total {
+                            flash_progress.total_size[operation] += total;
+                            flash_progress.size_done[operation] = 0;
```
I am confused why we are adding to `total_size` but resetting `size_done`.  I assume it is supposed to keep the sum of sizes if `AddProgressBar` is received multiple times (with `Some` size). I understand that this shouldn't happen after a `Progress` event, but `size_done = 0` would erase any progress if it did. Either way, since it already starts at zero, we can just remove this assignment?

arkeet replied:
Also, in case `total` is None, maybe that should be propagated to the `ProgressState` (to indicate the total size is unknown), and then pass `None` to `update_progress` (instead of possibly dividing by zero)?

bugadani replied:
There can be multiple erase-program phases, so total_size counts the total. OTOH vscode only diplays a single %counter at a time so resetting size_done is done so that a new phase starts from 0. But the 2 of them together indeed seems incorrect. Let's change this in a separate PR, though, but otherwise you're free to rethink this - it seems indeed incorrect.

Yatekii replied:
Tbh I designed this badly initially because of the way indicatif was designed and progressbars could not be reused. Imo we should only ever use one progressbar and have it completely independent from the library code. We should have a separate progress state for each phase and then just map the one progressbar to the currently active phase.

arkeet commented:
I have removed enum-map entirely, and switched to `HashMap`. I don't think anything was lost - there are no panicking operations and we still statically check that each `Operation` variant is handled. Also I changed `ProgressState` from a pair of maps to a map of pairs, since that just makes more sense.

bugadani reviewed (approved):
Thank you :)
